lib/rules/no-constant-binary-operand.js

```diff
@@ -0,0 +1,431 @@
+/**
+ * @fileoverview Rule to flag constant comparisons and logical expressions that always/never short circuit
+ * @author Jordan Eldredge <https://jordaneldredge.com>
+ */
+
+"use strict";
+
+const { isNullOrUndefined } = require("./utils/ast-utils");
+
+// Object keys which, if redefined, could impact how equality is computed.
+const OBJECT_EQUALITY_KEYS = new Set(["valueOf", "toString", "__proto__"]);
+const NUMERIC_OR_STRING_BINARY_OPERATORS = new Set(["+", "-", "*", "/", "%", "|", "^", "&"]);
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Test if an AST node has a statically knowable constant truthiness. Meaning,
+ * it will alwasy coerce to either `true` or `false`.
+ * @param {ASTNode} node The AST node being tested.
+ * @returns {boolean} Does `node` have constant truthiness?
+ */
+function hasConstantTruthiness(node) {
+    switch (node.type) {
+        case "ObjectExpression": // Objects are always truthy
+        case "ArrayExpression": // Arrays are always truthy
+        case "ArrowFunctionExpression": // Functions are always truthy
+        case "FunctionExpression": // Functions are always truthy
+        case "ClassExpression": // Classes are always truthy
+        case "NewExpression": // Objects are always truthy
+        case "Literal": // Truthy, or falsy, literals never change
+            return true;
+        case "CallExpression": {
+            if (node.callee.name === "Boolean") {
+                return node.arguments.length === 0 || hasConstantTruthiness(node.arguments[0]);
+            }
+            return false;
+        }
+        case "JSXElement": // ESLint has a policy of not assuming any specific JSX behavior.
+        case "JSXFragment":
+            return false;
+        case "AssignmentExpression":
+            if (node.operator !== "=") {
+                return false; // We won't go so far as to try to evaluate += etc.
+            }
+            return hasConstantTruthiness(node.right);
+        case "TemplateLiteral":
+
+            /*
+             * TODO: If all quasis are empty, we could look at node.expressions
+             * and try to determine if they are static truthinesss.
+             */
+            return node.quasis.some(quasi => quasi.value.cooked.length);
+        case "UnaryExpression":
+            if (node.operator === "void" || // Always returns `undefined`
+                node.operator === "typeof" // All type strings are truthy
+            ) {
+                return true;
+            }
+            if (node.operator === "!") {
+                return hasConstantTruthiness(node.argument);
+            }
+
+            /*
+             * We won't try to reason about +, -, ~, or delete
```

Comment:
`delete` can be easily reasoned about I think.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete#return_value

In non-strict mode it returns true unless the property being deleted is non-configurable.
In strict mode it always returns true and throws in the latter case.

So the logic could be:
```ts
if (context.getScope().isStrict) {
  // delete will always returns true or throw
  return true;
} else {
  // not strict mode
  if (node.type === 'MemberExpression') {
    // might return false if the property being deleted is non-configurable
    return false;
  }
  return true;
}
```

Reply:
For `+`, `-`, `~` - yeah they get a bit involved to handle - but it's possible to statically handle some cases.

Might be a good idea to write this as a TODO - in case anyone wants to solve this in future?

Reply:
`delete` is pretty rare, and supporting it will require passing scope (or at least `isStrict`) into the helpers. I think I'll punt on supporting `delete` for now, but if we end up with some other reason to include scope or other contextual data in the helpers, I'll add support at that point.